demos/programs/Exm/app_in_uil/exm_in_uil.man
.\" $XConsortium: exm_in_uil.man /main/4 1995/07/17 10:46:20 drk $
...\" @OPENGROUP_COPYRIGHT@
...\" COPYRIGHT NOTICE
...\" Copyright (c) 1990, 1991, 1992, 1993 Open Software Foundation, Inc.
...\" Copyright (c) 1996, 1997, 1998, 1999, 2000 The Open Group
...\" ALL RIGHTS RESERVED (MOTIF).  See the file named COPYRIGHT.MOTIF for
...\" the full copyright text.
...\" 
...\" This software is subject to an open license. It may only be
...\" used on, with or for operating systems which are themselves open
...\" source systems. You must contact The Open Group for a license
...\" allowing distribution and sublicensing of this software on, with,
...\" or for operating systems which are not Open Source programs.
...\" 
...\" See http://www.opengroup.org/openmotif/license for full
...\" details of the license agreement. Any use, reproduction, or
...\" distribution of the program constitutes recipient's acceptance of
...\" this agreement.
...\" 
...\" EXCEPT AS EXPRESSLY SET FORTH IN THIS AGREEMENT, THE PROGRAM IS
...\" PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
...\" KIND, EITHER EXPRESS OR IMPLIED INCLUDING, WITHOUT LIMITATION, ANY
...\" WARRANTIES OR CONDITIONS OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY
...\" OR FITNESS FOR A PARTICULAR PURPOSE
...\" 
...\" EXCEPT AS EXPRESSLY SET FORTH IN THIS AGREEMENT, NEITHER RECIPIENT
...\" NOR ANY CONTRIBUTORS SHALL HAVE ANY LIABILITY FOR ANY DIRECT,
...\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
...\" DAMAGES (INCLUDING WITHOUT LIMITATION LOST PROFITS), HOWEVER CAUSED
...\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
...\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
...\" ANY WAY OUT OF THE USE OR DISTRIBUTION OF THE PROGRAM OR THE
...\" EXERCISE OF ANY RIGHTS GRANTED HEREUNDER, EVEN IF ADVISED OF THE
...\" POSSIBILITY OF SUCH DAMAGES.
...\" 
...\" 
...\" HISTORY
.TH app_in_uil 1X MOTIF "Demonstration programs"
.SH NAME
\*Lapp_in_uil\*O\(emUIL application to exercise Exm demonstration widgets.
.SH SYNOPSIS
.sS
\*Lapp_in_uil\*O
.sE
.SH DESCRIPTION
\*Lapp_in_uil\*O
is a UIL application that exercises the Exm demonstration widgets.


NEWS
Wed May  7 08:35:18 EDT 2015
Mon Feb 22 16:20:59 EST 2021

TOP NEWS

  * for freeBSD 12.2 with Xorg installed sans Xprint (not in Xorg in 12.2)

  * 64 bit build, remade simple 'weatherable ' Makefiles, just several seconds
    to make + install with demos.

  * panner now has zoom in/out, trimmed down, stability improved

  * also has imake cf build working w/quats outside of X tree, but suggest
    as not mainstream (not weatherable, tends to break due to OS changes)

  * also "read to install" for freeBSD 12.2 amd64 (make install only)

---------
BELOW

  INSTALL
  NEWS
  New optional panner mwmrc Xdefault settings
  Why use mwm?
  CHANGELOG
  KNOWN compile warnings
  What is non-virtual panner?

---------
  INSTALL:
	$ make
	$ make install        # as root
	$ make demos          # not suggested as useful
  
  INSTALL:  imake  config/cf/Imake.tmpl style
	$ sh ./motif250.sh    # fixes stuff, sets env vars, imake, make
	$ make install        # as root
	$ make install.man    # as root

  note: if tests/ is in Imakefile quats will build too, see that file.
	quats are not automatic to use, are unwieldy, but build and work

---------
NEWS

openMotif-2.1.30
openMotif-2.2.3  (nearly same as 2.1.30)
openMotif-2.3.4  (some fixes from)
openMotif-2.2.5  (per say, but name is 2.5.0, see version naming below)

  Panner support is back, and is optional (can un-#define)

  Panner support is back, a non-vurtual compatible panner with added feature
    (OpenGroup deleted previous Virtual Panning before public release 2.1+)
    See: demos/programs/panner/panner.man

  * NEW panner can now Zoom in / out and resize all work together
  * NEW panner now more stable and trim than first release
  * panner(1) does edge panning (pan by mouse to edge of screen)
    edge panning did not need to be in mwm and is non-virtual
  * panner updates smoother without flashing it can animate (server side draw)
  * mwm panner support sees/pans all trasients (popups)
    and icons (although icon policy may change situationally)
  * this release is intented to be "a more weather resisant release"
    for using mwm as a window manager, meaning no cf (yet).
    see notes on why cf is fragile in issues section.  it is NOT aimed
    at providing a developer base.   See more below.
  * 2.2.3 redhat bug slow help open and wrong size problem fixed.
    it was a Xt 5 sec waiting on ICCC call due to hacked defaults.
    Now default is "yes support".  Still off by default in many
    automake Motif releases.  "polling" is also eliminated.

  from 2.1.30 to 2.3.4 Motif has barely changed except in lib/Xm lib/Mrm

  2.2.3 (thus 2.5.0) does not include the newer jpeg+fontconfig+international
  support from 2.3.4 libXm, thus also has less "lib depends headaches"

  ./localdef.h is added for making certain defaults easy to edit.  if ADD_PAN
    is Not defined: defines reverts to verbatim motif-2.0.2 (approx. 2.3.4)
  ./config.h is still around, side effect of autoconfig

---------
New mwmrc settings and .Xdefaults settings

By default panner is "active" and uses Alt-Arrow.  Settings are optional.
The panner demo has been re-vived and can drag windows in a view with mouse,
also for remote desktops.

FILE: .Xdefaults

Mwm*panScrollX : 	25
Mwm*panScrollY : 	25
Mwm*usePan :		True
Mwm*panUseX :		False  ! try True
Mwm*panUseMwm :		False

! default is use Mwm if pcd win, X if not
! useMwm uses mwm or nothin, useX uses X on all but does use Mwm
! useX should do all Mwm things w/o un-necessary code like resize checking

FILE: .mwmrc
! pan desktop by panScrollXY ammount

Keys DefaultKeyBindings
{
        Alt<Key>Up              root|window|icon        f.north
        Alt<Key>Down            root|window|icon        f.south
        Alt<Key>Right           root|window|icon        f.east
        Alt<Key>Left            root|window|icon        f.west
}


  About cf/
    They are too old to be useful on freeBSD 12.2
    The autoconfig release can also do imake build on linux
    using cf Motif-X11.cf.tar.gz (using X11R6 installed not Xorg)



--------------
Why use mwm?

  there's probably a book on it

  * you need lightweight yet reliable yet better than twm, or for nastalgia
  * you have lib issues and need somethig not tied into BIG TECH desktops

  Mwm:
	* well made and tested
	* transparent remote desktop support if needed
	* Motif has no lib dependancies outside of X11 ... better than
          twm and seems to work when all else is broken or not installed
	* (free) gui development toolkits for smart ui layout / application
	  making for Motif (like java ide or xcode but not feature competitive)
	* Motif has free + in-print books if you need help
	* some older apps work nicer (or dont crash or fub) using Motif
	* easier security (yet not as secure as twm)
	* good for developers - who don't want desk clutter and do use xterm

  Caveats:
	* not easy to get a point/click developer environment that works
	  with motif - though some old ones may exist
	* not android emulating, though maybe one runs in a window?
	* CDE desktop and Motif work together as a pair (not this 250 release)

---------
CHANGELOG

  * tools/mwl clients/uil reworked so that changes to motif.wml "really work"
    (automake distros use stale files and work only if never changed)
    and a minor fix to add a ';' in Uil.y knocked out by some past patch
    'HaL fix CR 5427' incorreclty applied, is my guess.

  * panner more stable

  * this is a freeBSD 12.2 release for Xorg so it is by default without
    DtPrint as Xorg canceled libXp for freeBSD (#undef NO_XP in localdef.h
    to enable -lXp, move directory wml.print/ to wml/ because motif.wml,
    should work).  for config cf build also edit config/cf/host.def NO_XP.

  * many very minor compilation changes, and spaces and ifdefs
    most clang warnings handled, a few checked ones skipped

  * the linux motif-2.5.0 release on www.sourceforge.org, but it is only
    for Xfree86-4.8.0 and uil "can't be modified" but ... The pan250

demos/programs/Exm/app_in_uil/exm_in_uil.c
/* $XConsortium: exm_in_uil.c /main/5 1995/07/15 20:44:21 drk $ */
/*
 * @OPENGROUP_COPYRIGHT@
 * COPYRIGHT NOTICE
 * Copyright (c) 1990, 1991, 1992, 1993 Open Software Foundation, Inc.
 * Copyright (c) 1996, 1997, 1998, 1999, 2000 The Open Group
 * ALL RIGHTS RESERVED (MOTIF).  See the file named COPYRIGHT.MOTIF for
 * the full copyright text.
 * 
 * This software is subject to an open license. It may only be
 * used on, with or for operating systems which are themselves open
 * source systems. You must contact The Open Group for a license
 * allowing distribution and sublicensing of this software on, with,
 * or for operating systems which are not Open Source programs.
 * 
 * See http://www.opengroup.org/openmotif/license for full
 * details of the license agreement. Any use, reproduction, or
 * distribution of the program constitutes recipient's acceptance of
 * this agreement.
 * 
 * EXCEPT AS EXPRESSLY SET FORTH IN THIS AGREEMENT, THE PROGRAM IS
 * PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, EITHER EXPRESS OR IMPLIED INCLUDING, WITHOUT LIMITATION, ANY
 * WARRANTIES OR CONDITIONS OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY
 * OR FITNESS FOR A PARTICULAR PURPOSE
 * 
 * EXCEPT AS EXPRESSLY SET FORTH IN THIS AGREEMENT, NEITHER RECIPIENT
 * NOR ANY CONTRIBUTORS SHALL HAVE ANY LIABILITY FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING WITHOUT LIMITATION LOST PROFITS), HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OR DISTRIBUTION OF THE PROGRAM OR THE
 * EXERCISE OF ANY RIGHTS GRANTED HEREUNDER, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGES.
 * 
 */
/*
 * HISTORY
 */

/*******************************************************************************
 *
 * exm_in_uil.c - main routine and C callbacks for exm_in_uil program.
 *
 ******************************************************************************/

#include <stdio.h>
#include <Xm/Xm.h>
#include <Xm/MessageB.h>
#include <Mrm/MrmPublic.h>
#include <Exm/ExmMrm.h>

static void ManageCb(Widget, String, XtPointer);
static void ExitCb(Widget, XtPointer, XtPointer);

static MrmHierarchy mrmId;
static MrmCode mrmClass;
static char *mrmFiles[]={"exm_in_uil.uid"};
static MRMRegisterArg mrmNames[] = {
        {"ManageCb", (caddr_t)ManageCb },
        {"ExitCb", (caddr_t)ExitCb }
};

static Display *display;
static Widget shell;

int 
main(int argc,
     char **argv)
{
Widget appMain;
XtAppContext appContext;

    MrmInitialize();
    ExmMrmInitialize();	/* register Exm widgets with Mrm */
    XtToolkitInitialize();
    appContext = XtCreateApplicationContext();
    display = XtOpenDisplay(appContext, NULL, "exmdemo", "ExmDemo", NULL, 0,
                            &argc, argv);
    if (display == NULL) {
	fprintf(stderr, "Can't open display\n");
	exit(1);
    }
    shell = XtAppCreateShell("demo", NULL, applicationShellWidgetClass,
						display, NULL, 0);
    if (MrmOpenHierarchy (1, mrmFiles, NULL, &mrmId) != MrmSUCCESS) 
      exit(1);
    MrmRegisterNames(mrmNames, XtNumber(mrmNames));
    MrmFetchWidget (mrmId, "appMain", shell, &appMain, &mrmClass);
    XtManageChild(appMain);
    XtRealizeWidget(shell);
    XtAppMainLoop(appContext);
    exit(0);
}



/*
 * ManageCb - Manage widget passed as client_data.
 */
static void 
ManageCb(Widget w,
	 String id,
	 XtPointer cb)
{
    XtManageChild (XtNameToWidget(shell, id));
}


/*
 * ExitCb - Exit application.
 */
static void 
ExitCb(Widget w,
       XtPointer cd,
       XtPointer cb)
{
    exit(0);
}
